src/devices/Ak8963/Ak8963.cs

```diff
@@ -0,0 +1,336 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Device.I2c;
+using System.IO;
+using System.Numerics;
+using System.Threading;
+
+namespace Iot.Device.Magnetometer
+{
+    /// <summary>
+    /// AK8963 class implementing a magnetometer
+    /// </summary>
+    public sealed class Ak8963 : IDisposable
+    {
+        private I2cDevice _i2cDevice;
+        private MeasurementMode _measurementMode;
+        private OutputBitMode _outputBitMode;
+        private bool _selfTest = false;
+        private Ak8963I2cBase _ak8963Interface;
+        private bool _autodispose = true;
+
+        /// <summary>
+        /// Default I2C address for the AK8963
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x0C;
+
+        /// <summary>
+        /// Default constructor for an independent AK8963
+        /// </summary>
+        /// <param name="i2CDevice">The I2C device</param>
+        public Ak8963(I2cDevice i2CDevice) : this(i2CDevice, new Ak8963I2c())
+        { }
+
+        /// <summary>
+        /// Constructor to use if AK8963 is behind another element and need a special I2C protocol like
+        /// when used with the MPU9250
+        /// </summary>
+        /// <param name="i2CDevice">The I2C device</param>
+        /// <param name="ak8963Interface">The specific interface to communicate with the AK8963</param>
+        /// <param name="autoDispose">True to dispose the I2C device when class is disposed</param>
+        public Ak8963(I2cDevice i2CDevice, Ak8963I2cBase ak8963Interface, bool autoDispose = true)
+        {
+            _i2cDevice = i2CDevice;
+            _ak8963Interface = ak8963Interface;
+            // Initialize the default modes
+            _measurementMode = MeasurementMode.PowerDown;
+            _outputBitMode = OutputBitMode.Output14bit;
+            _autodispose = autoDispose;
+            byte mode = (byte)((byte)_measurementMode | ((byte)_outputBitMode << 4));
+            WriteRegister(Register.CNTL, mode);
+            if (!CheckVersion())
```

Comment:
CheckVersion should not return and throw instead (or name should be changed to something starting with `Is...`, i.e. `IsVersionCorrect`)